### ISSUE TITLE
BuildConfig `secrets` not available during build

### ISSUE THREAD

**Issue by matthewvalimaki** [openshift/origin#17892]
`secrets` in `BuildConfig` work on build `A` but do not work with build `B` which is a copy of `A` with minor changes. The minor change is merely in the `dockerStrategy` pointing back to `A` in `kind: "ImageStreamTag"`. No matter which secrets I want to expose in build `B` they do not become available. The builder pod does show them as mounts.

I am wondering if build `A` using secrets somehow blocks secrets in the `B`, which again is using `A` as `kind: "ImageStreamTag"`. I have tried changing `destinationDir` but it does not seem to make a difference.

##### Version
```
oc v3.6.1+008f2d5
kubernetes v1.6.1+5115d708d7
```

##### Steps To Reproduce
1. Create two builds with second builds `ImageStreamTag` pointing to the first
2. Expose secrets in both builds
3. Observe that none of the secrets are in the second build

##### Current Result
Secrets are not available during build.

##### Expected Result
Any and all secrets should be accessible during build. In the example yamls below I am expecting to see `ssh-key` secret in `var/www/html/`.

##### Additional Information
Build `A`
```
apiVersion: v1
kind: BuildConfig
metadata:
  name: a
  namespace: development
spec:
  triggers:
    - imageChange:
      from:
        kind: "ImageStreamTag"
        name: "base:latest"
      type: ImageChange
  output:
    to:
      kind: "ImageStreamTag"
      name: "a:latest"
  strategy:
    dockerStrategy:
      noCache: true
      from:
        kind: "ImageStreamTag"
        name: "base:latest"
      dockerfilePath: Dockerfile
  source:
    secrets:
      - secret:
          name: ssh-key
          destinationDir: var/www/html
    sourceSecret:
      name: ssh-key
    git:
      uri: "git@git-server:development/project.git" 
```

Build `B`
```
apiVersion: v1
kind: BuildConfig
metadata:
  name: b
  namespace: development
spec:
  triggers:
    - imageChange:
      from:
        kind: "ImageStreamTag"
        name: "a:latest"
      type: ImageChange
  output:
    to:
      kind: "ImageStreamTag"
      name: "b:latest"
  strategy:
    dockerStrategy:
      noCache: true
      from:
        kind: "ImageStreamTag"
        name: "a:latest"
      dockerfilePath: Dockerfile
  source:
    secrets:
      - secret:
          name: ssh-key
          destinationDir: var/www/html
    sourceSecret:
      name: ssh-key
    git:
      uri: "git@git-server:development/additional-things.git" 
```

**Comment by php-coder** [comment]
@openshift/sig-developer-experience

**Comment by jim-minter** [comment]
@matthewvalimaki looking at the code comments (this is underdocumented, we'll deal with that...):

```
	// DestinationDir is the directory where the files from the secret should be
	// available for the build time.
	// For the Source build strategy, these will be injected into a container
	// where the assemble script runs. Later, when the script finishes, all files
	// injected will be truncated to zero length.
	// For the Docker build strategy, these will be copied into the build
	// directory, where the Dockerfile is located, so users can ADD or COPY them
	// during docker build.
```

For the Docker build B, is Origin making available the secret files so that you can use ADD or COPY on them in the Dockerfile?  This appears to be the intended behaviour.

**Comment by matthewvalimaki** [comment]
@jim-minter I have not tried ADD or COPY. I can certainly try but I am unsure of the logic. WORKDIR changes the context and with build A the secret was available in what the WORKDIR was. In above examples I removed WORKDIR and tried destinationDir instead. But perhaps they just are available in the build directory. So I would do `ADD ssh-privatekey /var/www/html`?

**Comment by php-coder** [comment]
> So I would do ADD ssh-privatekey /var/www/html?

Please, don't add any sensitive data (like private keys) to the image. The data will be committed to the image and can be extracted by any user who can access to this image.

**Comment by jim-minter** [comment]
@matthewvalimaki the best example I've found so far is our test case.  See the Dockerfile at https://github.com/openshift/origin/blob/master/test/extended/testdata/builds/build-secrets/Dockerfile and BuildConfig at https://github.com/openshift/origin/blob/master/test/extended/testdata/builds/build-secrets/test-docker-build.json .

**Comment by bparees** [comment]
> So I would do ADD ssh-privatekey /var/www/html?

yes, in your first build.  As @php-coder noted, this means your ssh-privatekey will be in the output image though, which obviously you need to evaluate against your security requirements for the image.

the way to use this is documented here:
https://docs.openshift.org/latest/dev_guide/builds/build_inputs.html#using-secrets-docker-strategy

```
When using a Docker strategy, you can add all defined input secrets into your container image using the ADD and COPY instructions in your Dockerfile.
```

and @php-coder's concerns are covered as well:
```
Users should always remove their input secrets from the final application image so that the secrets are not present in the container running from that image. However, the secrets will still exist in the image itself in the layer where they were added. This removal should be part of the Dockerfile itself.
```

@jim-minter if you want to add an explicit Dockerfile example to the above product doc, i'm sure that would help people out.

**Comment by bparees** [comment]
> @jim-minter if you want to add an explicit Dockerfile example to the above product doc, i'm sure that would help people out.

(as a way of closing this issue)

**Comment by matthewvalimaki** [comment]
Thank you all for the help. I just tested `COPY ./ssh-privatekey /var/www/html` and it indeed now works in build B. Now I have my secret where I want it to be. While improving documentation is a must I do want to point out that for some reason `ADD` or `COPY` is not required with build A. For some reason `ssh-privatekey` is in the whatever `WORKDIR` is. It would be interesting to know why but I am happy enough to proceed without explanation.

Perhaps off-topic but in the OpenShift Web Console it reports build container mount path for mentioned secret as `/var/run/secrets/openshift.io/build/ssh-key` yet there is no `/var/run/secrets/openshift.io/` at all. Before I go ahead and log something with the console project I was wondering if this is indeed expected.

**Comment by bparees** [comment]
>  I do want to point out that for some reason ADD or COPY is not required with build A. For some reason ssh-privatekey is in the whatever WORKDIR is. 

I don't know what you mean when you say it is "not required".  There are two steps required to consume a secret:

1) define in the buildconfig that you want the secret.  This will result in the secret being mounted into your build pod and made available in the docker working dir that your Dockerfile is run against
2) define in your Dockerfile that you want to add/copy the secret into the image you are building.

If you can share your Dockerfile and what you are observing in build A, perhaps that would help, but there is no way the secret ends up in your build A output image unless your Dockerfile takes an action to make that happen (perhaps your build A base image includes an ONBUILD instruction which is doing an ADD?).

And if that secret is ending up in your build A output image, then I don't know why you are adding it again in your build B which extends the build A output image.

Loglevel 5 output from each of your builds would show us exactly what docker operations are being performed during your build so you can see what is ADD/COPYed

> Perhaps off-topic but in the OpenShift Web Console it reports build container mount path for mentioned secret as /var/run/secrets/openshift.io/build/ssh-key yet there is no /var/run/secrets/openshift.io/ at all. 

There is no /var/run/secrets/openshift.io where?  What are you doing/looking at when you say it isn't there?

There are a number of moving pieces to builds and it's not always obvious to users what they are.  I suspect you are being confused by that.  The build pod will have /var/run/secrets/openshift.io/build/ssh-key mounted into it, but that's not the "working dir" for the docker build, so you cannot reference that path in your Dockerfile.

**Comment by matthewvalimaki** [comment]
@bparees thank you for the thoughtful reply.

> (perhaps your build A base image includes an ONBUILD instruction which is doing an ADD?).

Ha! Indeed that is what I had. I indeed have `ADD . /var/www/` in Dockerfile for build A while I did not have that in B. I typically do not do `ADD .` but instead `ADD root ` to separate build related matters from files that I need in the image. So this definitely explain why it works in A but not in B.

> There are a number of moving pieces to builds and it's not always obvious to users what they are. I suspect you are being confused by that.

Indeed I was getting confused. I was looking for `/var/run/secrets/openshift.io` in the image being built not realizing, like you pointed out, the mounts are for the `CONTAINER: DOCKER-BUILD`.

All my questions have been answered. Thank you for great support.

**Comment by bparees** [comment]
> All my questions have been answered. Thank you for great support.

you're welcome, glad we could help.  We'll update the doc to include an example Dockerfile that shows the necessary ADD command.

**Comment by bparees** [comment]
doc update: 
https://github.com/openshift/openshift-docs/pull/7214

**Comment by andyfeller** [comment]
@bparees : realize this issue is a little old, but is there an example of a Dockerfile that actually cleans up input secrets in a secure, trustworthy way?

**Comment by adambkaplan** [comment]
@andyfeller in a Dockerfile, no unless the `imageOptimizationPolicy` is set to `SkipLayers` in an OpenShift build. A typical Dockerfile with docker, buildah, or podman will create a layer for each directive - same is true for a default OpenShift build. If one of your layers adds sensitive content via `ADD` or `COPY`, that content will be present in the layer and can be inspected.

Using the `SkipLayers` policy will let you remove sensitive content once it is no longer needed, as the resulting image only has one layer on top of the base image layers. This is at the expense of worse layer caching.

**Comment by andyfeller** [comment]
Thanks for the feedback @adambkaplan! 🙇 So my only concern with `SkipLayers` is in regards to the build failing at a given instruction before any final cleanup instruction can be given.

I doubly ran into this issue as I was looking to install pip modules from a private pypi repository as well as ansible-galaxy roles from scm, which both require some secret to authenticate to run.  Most of the Dockerfile examples I see online take the easy route of supporting unauthenticated versions of these activities within the Dockerfile with more enterprise concerns left unmentioned.  I suppose the expectation is move these out of the Dockerfile with your CI process do the work instead.

**Comment by bparees** [comment]
> So my only concern with SkipLayers is in regards to the build failing at a given instruction before any final cleanup instruction can be given.

if that were to happen the build would fail and the image would not be pushed anywhere (no image would even be produced, since execution of the dockerfile would have failed).

Unless of course you do something like "RUN rm -rf secretstuff && true"

but if the docker instruction fails, anyway, the build will fail.

**Comment by andyfeller** [comment]
Any opinions about using a multi-stage build with non-final image handling secrets with results copied into final to being more secure?

**Comment by bparees** [comment]
> Any opinions about using a multi-stage build with non-final image handling secrets with results copied into final to being more secure?

definitely a reasonable technique, though using layer squashing would be simpler/more efficient.

**Comment by andyfeller** [comment]
Looking under the covers, it seems [multistage Docker-based builds will enable `SkipLayers` in v3.11 if not configured](https://github.com/openshift/origin/blob/v3.11.0/pkg/build/builder/docker.go#L131), so will give that a go.  Thanks @bparees @adambkaplan 🙇 

![why not both](https://media.giphy.com/media/3o85xIO33l7RlmLR4I/giphy.gif)

**Comment by bparees** [comment]
@andyfeller yes that is accurate.  and in 4.x i believe you *always* get skiplayers behavior today (@adambkaplan can confirm, i have gotten this backwards before), but we'll be making it possible to choose which behavior you want, explicitly, soon.

**Comment by adambkaplan** [comment]
@bparees @andyfeller Yes, you always get skip layers in 4.1, but this is more of a bug than anything else.
We fixed it for 4.2 with openshift/builder#35, defaulting to the normal layered behavior.
